### ISSUE TITLE
GTSRB ConvNet Challenge

### ISSUE THREAD

**Issue by fornoni** [fornoni/gtsrb#1]
### Pre-processing

Following the approach in [4], the input images are bi-linearly up/down-sampled to a fixed resolution of 32x32. The results are subsequently converted to gray scale by discarding the UV channels, in a YUV color space. The images are subsequently preprocessed, by globally whitening the Y-value distribution (make it have zero mean and unit variance), followed by a spatial contrastive normalization emphasizing edges.
### Network architecture

The network architecture is selected with efficiency constraints in mind, as it has to be trained/tested on a mid-range laptop CPU. A low-capacity network is thus favored, as it reduces both the computational costs and the risk of overfitting. To the same end, a dropout "layer" is employed after each convolutional stage and after the first hidden layer, as it greatly speeds up the training, while further preventing overfitting [6]. Dense connection tables are favored over sparse ones, as dropout already effectively reduces the number of active connections during training, while the formers are more efficiently implemented in torch7 using Matrix-Matrix multiplications (SpatialConvolutionMM).

In each layer, a rectifier activation function is chosen over more classical hyperbolic-tangent activation functions, as it favors sparsity [2], and max-pooling is consequently selected as it preserves the outputs of units with a low probability of activation (i.e. sparse feature maps) [1]. The number of convolutional kernels in each layer, as well as the number of hidden units were selected experimentally, sticking to the canonical "by-pyramid" approach [3] and keeping the efficiency constraints in mind.

The network used in the experiments is the following:

```
(1): nn.SpatialConvolutionMM(1 -> 16, 5x5)
(2): nn.ReLU
(3): nn.SpatialMaxPooling(2,2,2,2)
(4): nn.Dropout(0.400000)
(5): nn.SpatialConvolutionMM(16 -> 32, 5x5)
(6): nn.ReLU
(7): nn.SpatialMaxPooling(2,2,2,2)
(8): nn.Dropout(0.400000)
(9): nn.View
(10): nn.Linear(800 -> 64)
(11): nn.ReLU
(12): nn.Dropout(0.400000)
(13): nn.Linear(64 -> 43)
(14): nn.LogSoftMax

# of trainable parameters: 67307
```
### Using the ConvNet as a feature extractor

In order to evaluate the effectiveness of the representations learned by the model, the network is also used as a feature extractor, by stripping it of its last layers and using the generated outputs as features [5]. The extracted features are then fed as an input to a kNN classifier. 
### First results

The network was trained using fully stochastic SGD for about 20 epochs. The training/testing error rates and training/testing times are reported below.

```
+ train error rate = 6.97%      
+ time / sample = 2.11ms
+ total training time = 1921s (~32 minutes)

+ test error rate = 4.11%
+ testing time = 7.32s
+ time / sample = 0.58ms
```

The training error rate is higher than the testing error as it is measured online, during training.

Using the CNN learned with 20 epochs of training and stripping the output of the CNN after the hidden layer (and ReLU), we can use the CNN as a feature extractor. To evaluate the effectiveness of this approach, the obtained features maps are fed as an input to a kNN classifier (with k=1). The results obtained by using an earlier stage of the network as a feature extractor is also reported. All the results are compared against a baseline obtained by directly feeding the preprocessed images to the kNN classifier.

```
+ removing last 3 layers (using the output of the hidden layer, after rectification)
error rate: 4.29%

+ removing last 5 layers (using the output of the last convolutional layer)
error rate: 9.47%

+ directly using the preprocessed features
error rate: 37.09%
```

As it can be seen, results are consistent with [5](although the task and the CNNs are here much simpler), showing that the network has learned increasingly informative representations at each layer.
### Conclusions

A thorough experimental evaluation is necessary to properly compare different architectures and draw any conclusion. At the very least, a cross-validation procedure should be used to tune the hyper-parameters of each network. 

The multi-scale architecture proposed in [4] should definitely be considered, while additional computational resources (e.g training on multiple GPUs) would allow to test larger/deeper architectures.
### References

[1] Boureau, Y-Lan, Jean Ponce, and Yann LeCun. "A theoretical analysis of feature pooling in visual recognition." Proceedings of the 27th International Conference on Machine Learning (ICML-10). 2010.
[2] Glorot, Xavier, Antoine Bordes, and Yoshua Bengio. "Deep sparse rectifier neural networks." International Conference on Artificial Intelligence and Statistics. 2011.
[3] LeCun, Yann, et al. "Gradient-based learning applied to document recognition." Proceedings of the IEEE 86.11 (1998): 2278-2324.
[4] Sermanet, Pierre, and Yann LeCun. "Traffic sign recognition with multi-scale convolutional networks." Neural Networks (IJCNN), The 2011 International Joint Conference on. IEEE, 2011
[5] Ali Sharif Razavian, Hossein Azizpour, Josephine Sullivan, Stefan Carlsson. CNN Features off-the-shelf: an Astounding Baseline for Recognition. arXiv:1403.6382. 2014
[6] Srivastava, Nitish, et al. "Dropout: A simple way to prevent neural networks from overfitting." The Journal of Machine Learning Research 15.1 (2014): 1929-1958.

**Comment by fornoni** [comment]
### Update

Using a multi-scale architecture with the same number of convolutional stages and kernels (msNet), and tuning a bit the learning rate to converge faster, the training error drops to 3.32% after only 20 epochs of training. After about 50 epochs of training the error further drops to values between 2.6 and 2.9%.
